src/prefect/engine/task_runner.py

```diff
@@ -574,7 +574,9 @@ def check_task_is_cached(self, state: State, inputs: Dict[str, Result]) -> State
             - ENDRUN: if the task is not ready to run
         """
         if self.task.cache_for is not None:
-            candidate_states = prefect.context.caches.get(self.task.name, [])
+            candidate_states = prefect.context.caches.get(
```

Comment:
Something about the order of operations here feels slightly off to me. I think it's that the `context.caches` check is happening before the passed `state` is checked. By the time we reach this pipeline step, we already know the task is `Pending` and ready to run -- but it seems to me if it were explicitly provided with a `Cached` state, that should take precedence over the context cache. Too complex?

Reply:
That makes sense; will update.

Reply:
Updated w/ test